### DISCUSSION TITLE
hello from egypt [AI search engine that works with IIAB]

### DISCUSSION THREAD

**Discussion by Abdulrahman392011** [iiab/iiab#3796] [Ideas]
so I took the liberty to build a AI search engine that works with IIAB, hopefully it will be up to par.
the app basically lets you talk with the model normally while taking notes of things it need to search for. after the conversation is done it starts searching IIAB for relevant information and then it enlists the points that are related to the conversation and then it becomes dormant till you come back. keep in mind that the model have long-term not to preclude the short-term memory of the model. instruction for how to clear the model's long-term memory is included in case you're wondering.

[amalgamation.zip](https://github.com/user-attachments/files/16919421/amalgamation.zip)

**Comment by holta** [comment]
For those interested in @Abdulrahman392011's prior work, here's a bit of context:

- #3789
- #3790

**Comment by Abdulrahman392011** [comment]
[synchronizer.zip](https://github.com/user-attachments/files/17049185/synchronizer.zip)

here you go,    one tested and fully working proof of concept.     I tested it and it doesn't bug anymore.  it still could use a little tuning but it should work 85% as expected.

it remember what happens and initiate conversations about it if required. it also postpone most of the searching queries till after the conversation is finished in order to dedicate all of the compute power to the conversation and become as responsive as possible.

it neutralizes the thoughts that are timed after they are conversed about.

**Comment by Abdulrahman392011** [comment]
so I patched a few issues with it. this should work about 90% as expected.

[synchronizer_2.zip](https://github.com/user-attachments/files/18354003/synchronizer_2.zip)

**Reply by Abdulrahman392011** [reply]
sorry the previous upload was in developer mode, I uploaded it again after disabling developing prints, now it's functioning the way it should

**Reply by nzola** [reply]
> so I patched a few issues with it. this should work about 90% as expected.
> 
> [synchronizer_2.zip](https://github.com/user-attachments/files/18268669/synchronizer_2.zip)

Hi Abdul, There is no instructions how to install this AI option on IIAB. Can you provide one? I am interested to test it.

**Reply by Abdulrahman392011** [reply]
Unzip, there's txt file with the installation instructions

**Reply by nzola** [reply]
Ok. Thanks Abdul!

**Reply by nzola** [reply]
For the items to install thru iiab admin, I cannot see; wikihow_en_maxi_2023-03     47.7  gigabytes.
I have installed other 3 items, wikipedia, stackoverflow and Gutenberg

**Reply by holta** [reply]
@nzola FYI Kiwix no longer distributes wikiHow ZIM files since about a month ago.

**Reply by nzola** [reply]
You are right. I have Kiwihow on my former installs. Can I copy from another Box?
Why kiwi has stopped to distrbute kiwi how?

**Reply by Abdulrahman392011** [reply]
Carry on with the installation, it will not break on you. When it searches for wikihow it will find nothing but the code won't break.

I will modify the code so that it will diverge wikihow searches to the other websites.

**Reply by Abdulrahman392011** [reply]
[synchronizer_2_temp.zip](https://github.com/user-attachments/files/18353990/synchronizer_2_temp.zip)

this should work, but make sure that the date of the website packages you download matches the ones in the installation instructions. (this is a future bug) now it works, I checked.

**Reply by holta** [reply]
> You are right. I have [wikiHow] on my former installs. Can I copy from another Box?

Yes, you can definitely try:

https://wiki.iiab.io/go/FAQ#How_do_I_add_ZIM_files,_like_Wikipedia%3F

**Reply by nzola** [reply]
> Carry on with the installation, it will not break on you. When it searches for wikihow it will find nothing but the code won't break.
> 
> I will modify the code so that it will diverge wikihow searches to the other websites.

Ok. Great!

**Reply by nzola** [reply]
> > You are right. I have [wikiHow] on my former installs. Can I copy from another Box?
> 
> Yes, you can definitely try:
> 
> https://wiki.iiab.io/go/FAQ#How_do_I_add_ZIM_files,_like_Wikipedia%3F

Thanks Adam. I will give this a try as soon as the 3 items installation on my new iiab is completed.

**Reply by EMG70** [reply]
I have tried to follow the instructions from @Abdulrahman392011 but i seem to fail in the usage section
``` ########################################## usage of the program ########################################
open a terminal
cd into the app folder then paste:

python3 ./sync2.py

wait until the model is loaded. it will show you a message related to search and put you on a new line. 
then chat normally. 
########################################################################################################
```

**Reply by Abdulrahman392011** [reply]
I forget to add ollama on the pip installs, to fix this, you have to copy the following command to your terminal.

pip install ollama --break-system-package

I added it to the pip installs in the instructions and re-uploaded it, if you are to do it all over again, you won't need the command above. but just so you wouldn't need to re-install use the command.

**Reply by nzola** [reply]
> I have tried to follow the instructions from @Abdulrahman392011 but i seem to fail in the usage section
> 
> ```
> open a terminal
> cd into the app folder then paste:
> 
> python3 ./sync2.py
> 
> wait until the model is loaded. it will show you a message related to search and put you on a new line. 
> then chat normally. 
> ########################################################################################################
> ```

Should I ignore **python3 ./sync2.py**, but type this on terminal **pip install ollama --break-system-package**

**Reply by EMG70** [reply]
unzip synchronzer folder 
cd synchronizer
python3 ./sync2.py
If you get errors then follow @Abdulrahman392011 solution pip install ollama ......

**Reply by Abdulrahman392011** [reply]
@nzola It's a lot of downloads to eventually not get it to work. If you still having troubles then don't be shy to speak up. Do as @EMG70 said then tell us if you got it to work or not.

**Reply by nzola** [reply]
Ok

**Comment by EMG70** [comment]
Am i going into the correct app folder?
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/4c98c058

![Screenshot from 2024-12-31 17-51-22](https://github.com/user-attachments/assets/184c109f-d5db-4279-bb3f-801ad950ec2e)
![Screenshot from 2024-12-31 17-51-49](https://github.com/user-attachments/assets/9ee6dd5a-77ef-483b-acc6-d0af3be64370)

**Reply by Abdulrahman392011** [reply]
Let me check something, just a sec

**Comment by EMG70** [comment]
Looks working ok now,Thanks @Abdulrahman392011 
[screen-capture (1).webm](https://github.com/user-attachments/assets/d7ea8bf3-d273-4855-b9a1-e9dff0925a43)

**Comment by EMG70** [comment]
How do i search content from specific files eg gutenberg ?

**Reply by EMG70** [reply]
Please ignore above message,its all clear now as on screenshot.Very good project.
![Screenshot from 2024-12-31 18-29-04](https://github.com/user-attachments/assets/0208dc55-7b12-4a14-a8d8-fee45bf249a1)

**Reply by Abdulrahman392011** [reply]
There's no simple way to explain it. It takes notes as you're chatting for the stuff needs searching, after 10 minutes of inactivity it dedicate the compute to searching and it becomes unavailable for chatting. When it finds something it opens it in the browser as a tap and explains its reason why it thinks it is useful for you and you should go through the stuff it opened.

If you don't want to wait the 10 minutes period of inactivity then type (search now) without the brackets. 

Given that there's something from your conversation that it want to search for, it will start searching immediately.

But i remind you again that it will be unavailable for chatting while searching.

**Reply by EMG70** [reply]
It responds very well.I will play around it to get specific topics from a known book.
Warning,a bit long video.
[screen-capture (2).webm](https://github.com/user-attachments/assets/70c82e12-13f2-429c-a1d4-ab051ca5ccde)

**Comment by nzola** [comment]
I type; search now gutenberg, but nothing happens
admin@box:/home/synchronizer_2 $ python3 ./sync2.py


type < search now > to initiate search immediately, automated search will start after ten minutes of inactivity.
      keep this window opened if you would like the AI model to initiate conversations on it's own

############################################## available ###############################################


search now gutenberg
################################################ busy ##################################################

############################################## available ###############################################

**Reply by Abdulrahman392011** [reply]
Type search for something, then it will reply by what it knows without searching. 

After about 10 minutes it will start searching the library (Guttenberg, stackoverflow and Wikipedia)

If you want to skip the 10 minutes type search now

**Reply by Abdulrahman392011** [reply]
You guys are the first to install it, so tell me if it doesn't work. 

It works on my laptop, so if it doesn't work on your machine, then I must have missed something in the installation instructions.

Leave me a reply and I will try to be as responsive as possible and find the mistake and correct it.

**Reply by EMG70** [reply]
@nzola ,Please get in touch when you're free.Mine worked OK and I could search topics inside books.

**Reply by nzola** [reply]
> @nzola ,Please get in touch when you're free.Mine worked OK and I could search topics inside books.

Ok Ed. I will get in touch with you tonight. Just to let you know, I am using Raspberry 4 thru ssh, not GUI

**Reply by Abdulrahman392011** [reply]
It won't work on a raspberry pi. 

It need something stronger. It can work on the new nvidia jetson super development kit (250$). 

If you ran it on a laptop it will work. But laptops are expensive to deploy in numbers.

Sorry for the bad news

**Reply by EMG70** [reply]
@nzola ,Here's a reminder from the installation-instruction text.
![Screenshot From 2025-01-08 19-39-00](https://github.com/user-attachments/assets/399fc7b7-4cf5-4e97-a404-e99fff860a5e)

**Reply by nzola** [reply]
> @nzola ,Here's a reminder from the installation-instruction text. ![Screenshot From 2025-01-08 19-39-00](https://private-user-images.githubusercontent.com/52469992/401284447-399fc7b7-4cf5-4e97-a404-e99fff860a5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNzE2MTEsIm5iZiI6MTczNjM3MTMxMSwicGF0aCI6Ii81MjQ2OTk5Mi80MDEyODQ0NDctMzk5ZmM3YjctNGNmNS00ZTk3LWE0MDQtZTk5ZmZmODYwYTVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDIxMjE1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4YjMyODI2NzQ5YzBjZDFlNzYzMGViMjg4OTA0YTQ1OWY2MTdkNTVkZmM4OWFmMjEwNzAyMjUwZmQ3NmI3NjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ju4oJvqvOaJpOrL7RA1jrLHNvmH0YByErPBmubSVdtY)

Hi @EMG70 , @Abdulrahman392011 says it won't work on Raspberrypi. i

**Reply by Abdulrahman392011** [reply]
I think there's some confusion on the minimum hardware requirements

8 gigabytes of vRAM or unified RAM
40 tops of gpu compute 

Anything lower than this it won't work and even if you got it to work on swap memory, the performance will be too slow and impractical.

**Reply by EMG70** [reply]
Thanks @Abdulrahman392011. I was also trying point Nzola to the requirements in your txt .

**Reply by Abdulrahman392011** [reply]
> Thanks @Abdulrahman392011. I was also trying point Nzola to the requirements in your txt .

I should have added the hardware requirements to the installation requirements to avoid this. 

I will add it and re-upload.

**Reply by nzola** [reply]
Thanks @Abdulrahman392011 and @EMG70

**Comment by Abdulrahman392011** [comment]
[synchronizer_3.zip](https://github.com/user-attachments/files/18446509/synchronizer_3.zip)

it operates 95% as expected.

**Comment by rimit123** [comment]
hi i m not able to handel this coding part but i also want in my iiab i can get ollama or any offline model can you please help me out is it possible we can get any precofif image that i can burn and it started wrkng for me please you assistant will help me out let me know soon please

**Reply by Abdulrahman392011** [reply]
hey man, what's up!

**Reply by Abdulrahman392011** [reply]
it's not as difficult as it looks. download the last link. I regularly update it, as I am improving and adding new capabilities to it.

**Reply by Abdulrahman392011** [reply]
I am with you, tell me what you doing step by step. start by downloading and decompressing the zip file, tell me when you done.

**Reply by rimit123** [reply]
thank you so much abdul assalwalekum sorry that getting you reply after three days i hope you are fine there so you give me the file which i downloaded but i m not able to understand any thing from coding part so how could you help me out

**Reply by Abdulrahman392011** [reply]
hey man, walakomllslam. no worries about the delayed reply. I suggest you enable the notification for you email app so you would get an email like me when a message is left for you on github. tell me you downloaded and unzipped the file. go inside you will find a folder named "usage and installation instruction" go in and you will find a txt file named "installation instructions.txt" open it and tell me afterwards.

**Reply by rimit123** [reply]
Shall we do this tomorrow its danm late night here so i will be available tomorrow while opening this file in my laptop and is it possible i can connect with you on whats app if possible then share your contact number please thanks for your always quick response

**Comment by Abdulrahman392011** [comment]
[module.zip](https://github.com/user-attachments/files/23920141/module.zip)

**Reply by holta** [reply]
@IoshuaCastillejos does the above work for you?

**Reply by Abdulrahman392011** [reply]
[module_2026.zip](https://github.com/user-attachments/files/24555905/module_2026.zip)

@holta 
i tested the module.zip file and it had a few bugs. i fixed them in the module_2026.zip file.

**Comment by observeroftime02** [comment]
I played around a bit with the sync2.py script, and the usage is not immediately obvious:

How do I "cleanly" exit the script once started? Ctrl + X or C or D didn't terminate it, neither did any combination of the various exit commands. I had to terminate the python process to get my shell back.

The .zim files mentioned in the installation instructions don't exist in their original form anymore (there's newer versions of them). The full gutenberg library is now 200gb, the instructions mention a 71gb file. Is it safe to assume that this will work with other versions of the .zim? I got wikipedia_en_all_maxi_2025-08.zim, sackoverflow.com_en_all_2023-11.zim, and an assortment of smaller gutenberg library files (gutenberg_en_lcc-c_2025-12, gutenberg_en_lcc-d_2025-12, etc.) as opposed to the "all" library.

When "Talking" with the model (for example "what is leprosy") I get a lot of information that points me into the world wide web, nothing that says "have a look on your server _here_". Is this normal? Can someone who has successfully used the model to search their server's content post a quick overview of the way you interact with this to pull up information? Is an internet connection required at all?

I would absolutely love a better way of searching my .zim files and this seems to be doing this, EXCEPT, I am too daft to understand how to properly use this but not daft enough to not be able to install this.

**Comment by Abdulrahman392011** [comment]
Hi @observeroftime02 . Use the module.zip file in my latest comment.

We are trying to implement it in IIAB and the module is a search module automatically detect the zim files you installed and search in them and give an answer. But it doesn't let you ask follow up questions like sync.py does.

Sync.py was a prototype and it only functions with the specified zim files and the specified versions.

The module will open the source page on your local server automatically when it uses one.

**Comment by Abdulrahman392011** [comment]
@observeroftime02 by the way if you're keen on running sync.py , you can edit the file in a text editor and ctrl+f search for the name of the original version of the zim file which you don't have and change it's name to the newer version that you downloaded. For instance wikipedia-2024 to wikipedia-2025 and so on. Just remember to change all the found matches in the file. As far as i recall that should work. It's been a couple of years since i last done any modifications to that file.

The module.zip file is the main critical elements of sync. I removed anything that isn't necessary so it would be easier to review from IIAB prospective. It's like 600 lines in comparison to sync which is 2000 lines.

**Comment by Abdulrahman392011** [comment]
[module_2026.zip](https://github.com/user-attachments/files/24556109/module_2026.zip)
[synchronizer_2_2026.zip](https://github.com/user-attachments/files/24556110/synchronizer_2_2026.zip)

@observeroftime02  I modified the sync2.py file so that it's easier for you to change the zim files versions and names. just double click on the sync2.py file and you will find a dictionary with the names of zim files change the names and it will work with the zim files you have. the modified sync2.py is in synchronizer_2_2026.zip .

also you will find model_name which you can change to whatever model you want to use. i mean if you want to use a stronger model you can. just make sure to pull the model from ollama first.

also another thing about the install. you can use the install file in module_2026.zip . it simplify things. you can also use the module_2026.zip to test that the system works. it had minor bugs i just fixed and it should run smoothly. 

if you ran into issues. i can modify the file, just tell me.

**Comment by Abdulrahman392011** [comment]
@observeroftime02 

sync2.py is an old script. i iterated over it multiple times. here is the newest version. the newer version uses structured output with a 1 billion mixture of expert model from IBM. the newer iteration is much faster and more stable than sync2.py . i wanted to tell you how the script works so you would understand the usage. it basically chat without doing retrieval augmented generation but it takes notes about things it wants to search for. after about 10 minutes of you not talking to it, it starts doing the retrieval augmented generation. but while it's chatting with you, if you give a direct order of search for something it tells you that it has taken a note and will do it after it's done talking with you, unless you tell it 'search now'. in that case it leaves you and start the retrieval augmented generation. this wasn't what used to happen in sync2.py . it was simply responding normally to the search command without the retrieval but it still took note to search for afterwords and will do the retrieval after 10 minutes of inactivity. this was very confusing. but it's fixed in the new iteration. 

another thing the new iteration does text to speech. it generate a sentence and then talk then generate the next sentence and so on. it's a technique that reduce latency as it doesn't wait for the entire response to be complete before initiating the text to speech model. at the same time allow you to interrupt the model midway while talking without have already wasted the compute and battery on generating the rest of the response. the text to speech model can consume a bit of compute but you can turn it off to make it faster and you'll just read, then pass the flag --disable_sound_for_performance while running the sync.py file. there is a print reminder of this flag everytime you run the script.

there is one more thing. the script automatically take notes of events from your chat meaning if you told it you will go to the doctor next Monday for instance it will remember and next Monday it will ask you about it if you opened the script and if you left the script running it will initiate a response without you prompting it. 

i looked into the zim files you downloaded and modified the script to run on these. so you won't need to do any modifications. just run the script and it should work. but the installation is now different. there is a file for the install that you can run to install the added dependencies needed by the new iteration. it might do something to your ollama install that you will need to run ollama serve manually outside the script. i needed to increase the nice level of ollama for the text to speech model to run with ollama without stressing the system and make it visibly lag. the script manually run ollama serve if not already running so you might need to run it manually if you needed it outside the script. if you just want everything to return back to normal, re-install ollama again and everything will return back to normal.  

this iteration is meant to be used with the following zims (wikipedia_en_all_maxi_2024-01) for casual knowledge ( gutenberg_en_all_2023-08 ) for history and specific knowledge", ( khanacademy_en_all_2023-03 ) videos for entertainment", ( stackoverflow.com_en_all_2023-11 ) for computer programming topics. i encourage you to download them or the new version of them and modify the file back to it's original values. reminding you again these are the current zim values (you already downloaded them):

wikipedia_en_all_maxi_2025-08
gutenberg_en_lcc-c_2025-12
gutenberg_en_lcc-d_2025-12
sackoverflow.com_en_all_2023-11


you will find them in a dictionary in the script called 'websites_available' just remember to NOT include the .zim in the name of the file. just add the name of the file without the extension.

one last point, when you're testing the system it might need to download something the very first time but after that just turn off the WiFi just to confirm that it doesn't need internet connectivity. when it does retrieval it will open the source page on your local server in the zim files so you can confirm the augmented generation with the source page.


github have some weird hoops and loops to upload anything larger than 25 megabytes or something like that, the zip file is 300 megabytes cause of the text to speech model included. i uploaded it to my google drive , made it viewable by anyone and shared the link: https://drive.google.com/file/d/11OOC4zEfDUcn1PWUcWy4O-AE3CQ-8frC/view?usp=sharing

thanks for stirring this up, me and Adam (IIAB creator @holta ) has been going back and forth on this for a long time now. so thanks for moving things and sorry for the delay of this reply, it was the holidays and too much was going on at the same time.
